### ISSUE TITLE
[Bug]: Stuttery playback of DVB-T stream

### ISSUE THREAD

**Issue by kasper93** [GPUOpen-LibrariesAndSDKs/AMF#380]
**Describe the bug**
Playback of pretty standard DVB-T full hd stream is not smooth.

**To Reproduce**
1. Open [st.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/11067684/st.zip) in PlaybackHW
2. Playback is not smooth

**Setup (please complete the following information):**
 - OS: Windows 10 22H2
 - Driver Version 23.2.1
 - GPU RX 5700 XT
 - Decoder

**Expected behavior**
Smooth playback

**Additional context**
The same happens on 7950X iGPU with ffmpeg based decoder. PlaybackHW sample defaults to dedicated GPU, so not tested in this application.

The issue is not limited to this sample file.

**Comment by kasper93** [comment]
> According to Vega analyzer the video stream in the file is seriously corrupted.

Yes, this is expected. This is raw transport dump from the air. It doesn't start nicely, there is no sync point. I explicitly didn't process the file to avoid any modifications. Just skip to the part where there it all data available (+29 frames) I can record longer sample or different channels, if that helps. Also the parameters of those streams didn't change for like over 10 years, and I have recordings from back then which also doesn't work, while working on literally any other hardware decoder that supports h264.

Another stuttering sample:
[01-07_16-33-14_VTM 4_Top Gear - S30 E03. Freddie Flintoff, Chris Harris en Paddy M.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/11082445/01-07_16-33-14_VTM.4_Top.Gear.-.S30.E03.Freddie.Flintoff.Chris.Harris.en.Paddy.M.zip)

Here is another problematic sample, with timing issues. Playbacks at 12.5 fps and has severe audio issue in `PlaybackHW` in ffmpeg it is just hitching. (don't worry about the name, the files is from elsewhere)
[Sample2_#2750.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/11082371/Sample2_.2750.zip)

**Comment by MikhailAMD** [comment]
In these streams timestamps on video and audio streams are far away from each other. You can add the tracing line below at the end of AMFFileDemuxerFFMPEGImpl::AMFOutputDemuxerImpl::QueryOutput() method in FileDemuxerFFMPEGImpl.cpp:

    AMFTraceInfo(AMF_FACILITY, L"Stream# %d, packets read %d pts=%lld", m_iIndexFFmpeg, (int)m_iPacketCount, buf->GetPts());

If one disables AV sync in presenters, the audio playback is OK. You can comment these lines in PlaybackPipelineBase.cpp

        m_pVideoPresenter->SetAVSyncObject(&m_AVSync);
        m_pAudioPresenter->SetAVSyncObject(&m_AVSync);

Note, that I am testing with the newer FFmpeg for demuxing, the update will be public with AMF SDK update. 
As you stated you see the same issue with FFmpeg. 
Note, that PlaybackHW is a sample app, it shows mostly how to use AMF components, it is not a full player.

**Comment by kasper93** [comment]
> In these streams timestamps on video and audio streams are far away from each other. You can add the tracing line below at the end of AMFFileDemuxerFFMPEGImpl::AMFOutputDemuxerImpl::QueryOutput() method in FileDemuxerFFMPEGImpl.cpp:

Indeed, but it shouldn't be a problem, no? Anyway, I don't care about the audio.

>  Note, that PlaybackHW is a sample app, it shows mostly how to use AMF components, it is not a full player.

To be perfectly transparent with you, I'm using PlaybackHW as an excuse to report the issue here, reach you guys, and have better chance to shed some light on the issue.

>  As you stated you see the same issue with FFmpeg.

Yes, indeed. It is easier to test with ffmpeg itself. With simple reencode with d3d11va (upstream ffmpeg doesn't implement AMF decoder https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/199 hence I'm using d3d11va api)
```
ffmpeg -hwaccel d3d11va -i "01-07_16-33-14_VTM 4_Top Gear - S30 E03. Freddie Flintoff, Chris Harris en Paddy M.ts" -an out.mkv
```
You can see that the `out.mkv` is corrupted. I also removed audio, as we don't need it.

How would you suggest we proceed on fixing this issue?

P.S. It seems to work ok on Linux with same hardware, so it makes me fell it is driver issue on Windows only.

**Comment by MikhailAMD** [comment]
I extracted elementary streams from the TS files with this CMD:
**ffmpeg.exe -i %1 -vcodec copy -bsf h264_mp4toannexb -an %1.h264** 
FFmeg complained about corruption and I believe skipped some frames. After that elementary stream 01-07_16-33-14_VTM 4_Top Gear - S30 E03. Freddie Flintoff, Chris Harris en Paddy M.ts.h264 is played without corruption and st.ts.h264 has one corruption in the middle, looking like a dropped frame. 
I tried PlaybackHW with "-swdecoder true" option. It forces to use FFmpeg SW decoder. The results are the same as with AMF decoder. 
At this point I don't see problem with AMF or VCN decoder, rather then file issues.

**Comment by MikhailAMD** [comment]
Seems not AMF issue. Please reopen if needed.

**Comment by kasper93** [comment]
> Seems not AMF issue. Please reopen if needed.

I don't have permissions to reopen. I can file a fresh issue if that is preferred. The issue is very much still present, now on RDNA4 (RX 9070 XT) with the 26.6.4 driver. I don't want to argue about the "correctness" of the files shared before, because that was never the main point. The main point is that VCN enters a corrupted state and never recovers, even after decoding crosses multiple keyframes and recovery points. It's natural that OTA transmissions like DVB-T, or abruptly cut MPEG-TS files like these, may feed the decoder invalid data, but it should recover, even if playback is briefly broken. Wouldn't you agree? There are multiple ways to trigger the corrupted state, and only a full decoder reinit clears it.

To make it easier to diagnose this time, I asked a friendly robot to create a standalone reproducer, so you can reproduce this directly against the decoder, without any player involved. Sorry for the wall of text, I know it may be hard to ingest, but I tried to keep it as concise as possible while staying coherent.

[st2.zip](https://github.com/user-attachments/files/30421371/st2.zip)

### TL;DR

On PAFF interlaced H.264, the standard DVB broadcast coding, the decoder permanently corrupts all inter prediction after either (a) re-entering a stream mid-GOP, e.g. any seek, or (b) being abandoned after the first field of a non-reference field pair, which a seek does with high probability. It never resynchronizes, across arbitrarily many recovery points, and decode time also rises several times (VCN is pinned and 100% usage when the bug happens). Only destroying the decoder instance recovers. Progressive and MBAFF interlaced content is not affected - which is why most files appear to work fine and this has been so easy to misattribute to "broken files".

### Setup

RX 9070 XT (VCN 5), Adrenalin 26.6.4 (31021.5001), Windows 11, FFmpeg master (705890061). Historically reproduced on everything from Vega to RDNA4 since driver 21.2.3.

### Sample

`st2.ts` (in the attached `st2.zip`) - a 13 s clean cut of the previously shared `st.ts`: 1080i25 PAFF field pictures, open GOP with recovery points every ~1.2 s (standard DVB coding). Software decode reports **zero** errors, and the hardware decoder itself decodes it **bit-exactly** when played linearly:

```
ffmpeg -v error -i st2.ts -f null -      (no output)
hwverify -i st2.ts -hw -verify -p D
# VERDICT: 0/308 frames differ from software decode: OK, bit-exact
```

### Reproducer

`hwverify.c` (in `st2.zip`, single file, plain FFmpeg API). It decodes with D3D11VA (the AMF decoder is affected too, runnable with `-c h264_amf`) while feeding the *identical packets* to the software decoder and compares outputs frame by frame (H.264 is bit-exact by spec). It prints a verdict, exits with code 3 when the decoder did not recover, and `-o` writes the hardware output as lossless FFV1 so you can watch the corruption.

### Trigger 1: seek into the middle of a GOP

What any player seek in mpegts does:

```
hwverify -i st2.ts -hw -verify -p d100,s6.3,D -o issue1_hw.mkv
# VERDICT: 131/240 frames differ from software decode (first=101 last=239 of 240):
#          decoder DID NOT RECOVER (still corrupt at end of stream)
```

### Trigger 2: abandoned first field of a non-reference field pair

Feed exactly 205 packets - the last one is the first field of a non-reference pair - then seek back, so its second field never arrives:

```
hwverify -i st2.ts -hw -t 1 -verify -p p205,s0,D -o issue2_hw.mkv
# VERDICT: 256/373 frames differ from software decode: decoder DID NOT RECOVER

hwverify -i st2.ts -hw -t 1 -verify -p p206,s0,D      (one packet more: pair completed)
# VERDICT: 0/374 frames differ from software decode: OK, bit-exact
```

A single packet separates a bit-exact run from a permanently corrupted one. A player seek stops the feed at an arbitrary access unit, so on typical broadcast content roughly every third seek hits this.

### Why this is inside the decoder

- We logged every DXVA submission (picture params with the full DPB view, all buffers, Begin/Execute/EndFrame results) for the corrupt and the clean run of trigger 2: after the seek both runs receive **byte-identical submissions (0 of 613 differ)**, each a complete BeginFrame..EndFrame cycle. The only difference the decoder ever saw is the one abandoned field. DXVA is a stateless interface - every subsequent frame carries a complete, consistent DPB description - yet the internal state never resynchronizes.
- After the trigger, intra pictures are still bit-exact indefinitely while ~97% of inter pictures are corrupt indefinitely, and per-frame decode time rises from ~1 ms to ~7 ms permanently. The corrupt frames mostly look plausible but contain content from the wrong point in time (wrong reference pictures) - watch the `issue*_hw.mkv` files that the commands above produce: playback visibly jumps between different moments.
- Destroying and recreating the decoder instance mid-stream fully restores correctness, even when the same texture array is reused.
- Only field-coded (PAFF) streams are affected: identical experiments on progressive and MBAFF encodes of the same content are bit-exact.
- The same permanent corruption reproduces through the AMF decoder (`h264_amf`), and a matching report exists for Linux VAAPI/Mesa (https://code.ffmpeg.org/FFmpeg/FFmpeg/issues/21290), so this looks like shared VCN firmware/core rather than a single driver frontend.

### Relation to FFmpeg's flush change

To avoid confusion: until 2025 FFmpeg recreated the hardware decoder session on every `avcodec_flush_buffers()`, and removing that (a legitimate optimization - DXVA is stateless by design) made the problem much more visible in FFmpeg-based players (see https://code.ffmpeg.org/FFmpeg/FFmpeg/issues/21290, https://github.com/mpv-player/mpv/issues/17136, https://github.com/Nevcairiel/LAVFilters/issues/695). To be clear, that change is *not* the root cause: this issue was filed in 2023, when the recreation was still in place, and seeking corrupted playback back then too. The decoder's internal state handling is the common factor across all these reports.

### Workarounds

Applications can partially work around this by tearing down and recreating the decoder on every seek *and* resuming decoding only at a keyframe - recreation alone is not sufficient, since the AMF decoder corrupts even from a cold start when the first fed data is mid-GOP. But nothing covers live OTA playback: a transmission dropout creates the same discontinuity with no seek involved (we verified that a data gap without any decoder flush poisons the state the same way). The application cannot prevent that; it can only detect the damage after the fact and recreate the decoder. This is why it needs fixing in the decoder itself - it should recover on its own, like every other implementation does (software, Intel, NVIDIA, pre-21.x AMD drivers).

Happy to provide the submission traces, more samples, or run any experiment you need.

**Comment by MikhailAMD** [comment]
Reopened as not fixed. As it is seem DXVA/D3D11VA decoder issue, I will notify corresponding team.

**Comment by kasper93** [comment]
> As it is seem DXVA/D3D11VA decoder issue, I will notify corresponding team.

Thanks. As mentioned the issues is also reproducible with `h264_amf` decoder, as it is exposed now in ffmpeg too. The corruption is different, but still testcase 1 breaks. I just use ffmpeg, because it's the easiest to work with. Either way, I appreciate any eyes looking at this issue. I'm sure eventually we will arrive at some solutions :)

**Comment by MikhailAMD** [comment]
> > As it is seem DXVA/D3D11VA decoder issue, I will notify corresponding team.
> 
> Thanks. As mentioned the issues is also reproducible with `h264_amf` decoder, as it is exposed now in ffmpeg too. The corruption is different, but still testcase 1 breaks. I just use ffmpeg, because it's the easiest to work with. Either way, I appreciate any eyes looking at this issue. I'm sure eventually we will arrive at some solutions :)

Sure.  On D3D11 code path AMF is using the same D3D11VA API. There are some differences designed to provide more features, but they are not relevant to this issue.  I've notified the team and will keep an eye.
